src/keystore/InviteStore.ts

```diff
@@ -0,0 +1,76 @@
+import { keystore } from '@xmtp/proto'
+import { Persistence } from './persistence/interface'
+import type { TopicData } from './interfaces'
+import { Mutex } from 'async-mutex'
+import { typeSafeTopicMap } from './utils'
+
+const INVITE_KEY = 'invitations:v1'
+
+/**
+ * InviteStore holds a simple map of topic -> TopicData and writes to the persistence layer on changes
+ */
+export default class InviteStore {
+  private persistence?: Persistence
+  private mutex: Mutex
+  private topicMap: Map<string, TopicData>
+
+  constructor(
+    persistence?: Persistence,
+    initialData: Map<string, TopicData> = new Map()
+  ) {
+    this.persistence = persistence
+    this.mutex = new Mutex()
+    this.topicMap = initialData
+  }
+
+  static async create(persistence?: Persistence): Promise<InviteStore> {
+    if (persistence) {
+      const rawData = await persistence.getItem(INVITE_KEY)
+      if (rawData) {
+        try {
+          const inviteMap = typeSafeTopicMap(keystore.TopicMap.decode(rawData))
+          // Create an InviteStore with data preloaded
+          return new InviteStore(
+            persistence,
+            new Map(Object.entries(inviteMap))
+          )
+        } catch (e) {
+          console.warn(`Error loading invites from store: ${e}`)
+        }
+      }
+    }
+    return new InviteStore(persistence)
+  }
+
+  async add(topicData: TopicData[]): Promise<void> {
+    await this.mutex.runExclusive(async () => {
```

Comment:
Afraid not. It lives in the process. It's only there to protect against multiple async calls to the `add` function.

Reply:
Definitely an option. It also has some performance benefits, since we don't have to serialize everything to perform updates.

The reason In shied away from that approach initially is just that it's a little awkward given `LocalStorage` is shared space. To do the initial data load, we'd have to get the full list of keys and perform some validation against them to see if they were "ours" or something totally unrelated. Obviously doable though.

Reply:
I looked for some LocalStorage benchmarks. According to this [stackoverflow](https://stackoverflow.com/a/59016859) and reproduced locally, single key large payload writes take ~1ms on my machine and ~3ms in the example.

Definitely possible to get race conditions, but shouldn't be very common. Bad behavior requires the user to have multiple tabs and for the writes to be different (add topic A, add topic B) such that one topic is lost.

I do think Chrome caches LocalStorage reads very aggressively which could be an argument for splitting into separate keys with semi-immutable values. A benchmark which reads the same value repeatedly is consistently fast even if the value size increases:
```
read local storage - single x 5,552,479 ops/sec ±0.36% (69 runs sampled)
read local storage - small x 5,419,302 ops/sec ±0.49% (67 runs sampled)
read local storage - large x 5,285,241 ops/sec ±0.40% (69 runs sampled)
```

Reply:
FWIW, this implementation only touches LocalStorage on startup and on update. Reads are from memory only.